includes/sanitizers/class-amp-core-theme-sanitizer.php

```diff
@@ -1522,4 +1569,446 @@ public function add_twentyfourteen_search() {
+					// however it seems like AMP has turned this into a common mechanism for
+					// focusing on inputs after a lightbox (i.e. modal) opens.
+					if ( $focus_element instanceof DOMElement && in_array( $focus_element->tagName, [ 'input', 'select', 'textarea' ], true ) ) {
+						$focus_element->setAttribute( 'autofocus', true );
+					}
+				}
+			}
+		}
+	}
+
+	/**
+	 * Get the ID for an element.
+	 *
+	 * If the element does not have an ID, create one first.
+	 *
+	 * @param DOMElement $element Element to get the ID for.
+	 * @param string     $prefix  Optional. Defaults to '_amp_wp_id_'.
+	 * @return string ID to use.
+	 */
+	protected function get_element_id( $element, $prefix = 'amp-wp-id' ) {
+		static $index_counter = [];
+
+		if ( $element->hasAttribute( 'id' ) ) {
+			return $element->getAttribute( 'id' );
+		}
+
+		if ( ! array_key_exists( $prefix, $index_counter ) ) {
+			$index_counter[ $prefix ] = 2;
+			$element->setAttribute( 'id', $prefix );
+
+			return $prefix;
+		}
+
+		$id = "{$prefix}-{$index_counter[ $prefix ]}";
+		$index_counter[ $prefix ] ++;
+
+		$element->setAttribute( 'id', $id );
+
+		return $id;
+	}
+
+	/**
+	 * Provides a "best guess" as to what XPath would mirror a given CSS
+	 * selector.
+	 *
+	 * This is a very simplistic conversion and will only work for very basic
+	 * CSS selectors. Therefore, it provides a filter for themes & plugins to
+	 * hook into to provide custom replacements.
+	 *
+	 * @param string $css_selector CSS selector to convert.
+	 * @return string|null XPath that closely mirrors the provided CSS
+	 *                     selector, or null if an error occurred.
+	 */
+	protected function xpath_from_css_selector( $css_selector ) {
+		// Start with basic clean-up.
+		$css_selector = trim( $css_selector );
+		$css_selector = preg_replace( '/\s+/', ' ', $css_selector );
+
+		$xpath             = '';
+		$direct_descendant = false;
+		$token             = strtok( $css_selector, ' ' );
+
+		while ( false !== $token ) {
+			$matches = [];
+
+			// Direct descendant.
+			if ( preg_match( '/^>$/', $token, $matches ) ) {
+				$direct_descendant = true;
+				$token             = strtok( ' ' );
+				continue;
+			}
+
+			// Single class.
+			if ( preg_match( '/^\.(?<class>[a-zA-Z0-9-_]*)$/', $token, $matches ) ) {
+				$descendant        = $direct_descendant ? '/' : '//';
+				$xpath            .= "{$descendant}*[ @class and contains( concat( ' ', normalize-space( @class ), ' ' ), ' {$matches['class']} ' ) ]";
+				$direct_descendant = false;
+				$token             = strtok( ' ' );
+				continue;
+			}
+
+			// Element.
+			if ( preg_match( '/^(?<element>[^.][a-zA-Z0-9-_]*)$/', $token, $matches ) ) {
+				$descendant        = $direct_descendant ? '/' : '//';
+				$xpath            .= "{$descendant}{$matches['element']}";
+				$direct_descendant = false;
+				$token             = strtok( ' ' );
+				continue;
+			}
+
+			$token = strtok( ' ' );
```

Comment:
Magic!

Reply:
... could you please "rephrase"? Not sure if this is good or bad here...

Reply:
Just that this is a special function that magically has internal state.